algorithms/src/snark/marlin/data_structures/proof.rs

```diff
@@ -131,99 +134,136 @@ impl<F: PrimeField> Evaluations<F> {
-            g_b_eval: CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?,
-            g_c_eval: CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?,
+            g_a_evals: deserialize_vec_without_len(&mut reader, compress, validate, batch_sizes.len())?,
+            g_b_evals: deserialize_vec_without_len(&mut reader, compress, validate, batch_sizes.len())?,
+            g_c_evals: deserialize_vec_without_len(&mut reader, compress, validate, batch_sizes.len())?,
         })
     }
 }
 
 impl<F: PrimeField> Evaluations<F> {
-    pub(crate) fn from_map(map: &std::collections::BTreeMap<String, F>, batch_size: usize) -> Self {
-        let z_b_evals = map.iter().filter_map(|(k, v)| k.starts_with("z_b_").then_some(*v)).collect::<Vec<_>>();
-        assert_eq!(z_b_evals.len(), batch_size);
-        Self { z_b_evals, g_1_eval: map["g_1"], g_a_eval: map["g_a"], g_b_eval: map["g_b"], g_c_eval: map["g_c"] }
+    pub(crate) fn from_map(
+        map: &std::collections::BTreeMap<String, F>,
+        batch_sizes: BTreeMap<CircuitId, usize>,
+    ) -> Self {
+        let mut z_b_evals_collect: BTreeMap<CircuitId, Vec<F>> = BTreeMap::new();
+        let mut g_a_evals = Vec::with_capacity(batch_sizes.len());
+        let mut g_b_evals = Vec::with_capacity(batch_sizes.len());
+        let mut g_c_evals = Vec::with_capacity(batch_sizes.len());
+
+        for (label, value) in map {
+            if label == "g_1" {
+                break;
+            }
+
+            let circuit_id = CircuitId::from_witness_label(label);
+            if label.contains("z_b_") {
+                if let Some(z_b_i) = z_b_evals_collect.get_mut(&circuit_id) {
+                    z_b_i.push(*value);
+                } else {
+                    let mut values = Vec::with_capacity(batch_sizes[&circuit_id]);
+                    values.push(*value);
+                    z_b_evals_collect.insert(circuit_id, values);
+                }
+            } else if label.contains("g_a") {
+                g_a_evals.push(*value);
+            } else if label.contains("g_b") {
+                g_b_evals.push(*value);
+            } else if label.contains("g_c") {
+                g_c_evals.push(*value);
+            }
+        }
+        let z_b_evals = z_b_evals_collect.into_values().collect();
+        Self { z_b_evals, g_1_eval: map["g_1"], g_a_evals, g_b_evals, g_c_evals }
     }
 
-    pub(crate) fn get(&self, label: &str) -> Option<F> {
-        if let Some(index) = label.strip_prefix("z_b_") {
-            self.z_b_evals.get(index.parse::<usize>().unwrap()).copied()
+    pub(crate) fn get(&self, circuit_index: usize, label: &str) -> Option<F> {
+        if label == "g_1" {
+            return Some(self.g_1_eval);
+        }
+
+        if let Some(index) = label.find("z_b_") {
+            let z_b_eval_circuit = &self.z_b_evals[circuit_index];
+            let instance_index = label[index + 4..].parse::<usize>().unwrap();
```

Comment:
should the size of this `usize` differ across architectures? if not, we should be specific about it by picking e.g. `u64`, as we're dealing with indexing

Reply:
See discussion below where we serialize into the usize.